File: python/examples/sensor_uplink.py
import time
import logging
from iot26_edge.client import IoT26Client

logging.basicConfig(level=logging.INFO)
log = logging.getLogger("sensor_uplink")

def main():
    # Initialize the Python IoT26 Client
    client = IoT26Client(
        device_id="your-device-uuid",
        device_token="your-device-token",
        mqtt_broker="localhost",
        mqtt_port=1883
    )

    # Listen for downlink commands from the widget
    def on_command(cmd: dict):
        log.info(f"Received command: {cmd.get('action')}")
        
        if cmd.get("action") == "custom":
            log.info("Processing custom action command...")
            
            # Publish an uplink telemetry point back to sync the UI!
            log.info("Publishing uplink to update the widget UI...")
            client.publish_readings([
                {"sensor_id": "valve-sensor-uuid", "value": 1.0, "unit": "state"}
            ])

    client.on_command = on_command

    client.connect()
    client.loop_start()

    # Periodically publish telemetry
    try:
        while True:
            log.info("Publishing routine telemetry...")
            client.publish_readings([
                {"sensor_id": "temp-sensor-uuid", "value": 24.5, "unit": "C"},
                {"sensor_id": "hum-sensor-uuid", "value": 60.0, "unit": "%"}
            ])
            time.sleep(10)
    except KeyboardInterrupt:
        pass
    finally:
        client.disconnect()

if __name__ == "__main__":
    main()

File: dart/example/sensor_uplink.dart
import 'dart:io';
import '../lib/iot26_client.dart'; // Import the local SDK wrapper

void main() async {
  final client = Iot26Client(Iot26Config(
    deviceId: 'your-device-uuid',
    deviceToken: 'your-device-token',
    broker: 'localhost',
    port: 1883,
  ));

  // Register command listener
  client.commands.listen((Iot26Command cmd) {
    print('Received command action: ${cmd.action}');

    if (cmd.action == 'custom') {
      print('Processing custom command...');

      // PUBLISH UPLINK: Report the new state back to the dashboard widget!
      print('Publishing uplink to sync widget UI...');
      client.publishReadings([
        Iot26Reading(sensorId: 'valve-sensor-uuid', value: 1.0, unit: 'state'),
      ]);
      print('Uplink published successfully!');
    }
  });

  // Connect to the broker
  try {
    await client.connect();
    print('Connected to IoT26 MQTT broker!');
  } catch (e) {
    print('Failed to connect: $e');
    exit(1);
  }

  // Simulate periodic routine telemetry
  while (true) {
    await Future.delayed(Duration(seconds: 10));
    print('Publishing routine telemetry...');
    client.publishReadings([
      Iot26Reading(sensorId: 'temp-sensor-uuid', value: 24.5, unit: 'C'),
      Iot26Reading(sensorId: 'hum-sensor-uuid', value: 60.0, unit: '%'),
    ]);
  }
}
